src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,241 @@
+            .iter()
+            .map(|f| f.to_string())
+            .collect_vec();
+
+        Self {
+            package_id,
+            package,
+            version,
+            features,
+            build_type,
+            extern_crate_name: dep.extern_crate_name.to_string(),
+            dependencies,
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct SbomTarget {
+    kind: TargetKind,
+    crate_type: Option<CrateType>,
+    name: String,
+    edition: String,
+}
+
+impl From<&Target> for SbomTarget {
+    fn from(target: &Target) -> Self {
+        SbomTarget {
+            kind: target.kind().clone(),
+            crate_type: target.kind().rustc_crate_types().first().cloned(),
+            name: target.name().to_string(),
+            edition: target.edition().to_string(),
+        }
+    }
+}
+
+#[derive(Serialize, Clone)]
+struct SbomRustc {
+    version: String,
+    wrapper: Option<PathBuf>,
+    commit_hash: Option<String>,
+    host: String,
+}
+
+impl From<&Rustc> for SbomRustc {
+    fn from(rustc: &Rustc) -> Self {
+        Self {
+            version: rustc.version.to_string(),
+            wrapper: rustc.wrapper.clone(),
+            commit_hash: rustc.commit_hash.clone(),
+            host: rustc.host.to_string(),
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct Sbom {
+    format_version: SbomFormatVersion<1>,
+    package_id: PackageIdSpec,
+    name: String,
+    version: String,
+    source: String,
+    target: SbomTarget,
+    profile: Profile,
+    packages: Vec<SbomPackage>,
+    features: Vec<String>,
+    rustc: SbomRustc,
+}
+
+impl Sbom {
+    pub fn new(unit: &Unit, packages: Vec<SbomPackage>, rustc: SbomRustc) -> Self {
+        let package_id = unit.pkg.summary().package_id().to_spec();
+        let name = unit.pkg.name().to_string();
+        let version = unit.pkg.version().to_string();
+        let source = unit.pkg.package_id().source_id().to_string();
+        let target = (&unit.target).into();
+        let profile = unit.profile.clone();
+        let features = unit.features.iter().map(|f| f.to_string()).collect();
+
+        Self {
+            format_version: SbomFormatVersion,
+            package_id,
+            name,
+            version,
+            source,
+            target,
+            profile,
+            packages,
+            features,
+            rustc,
+        }
+    }
+}
+
+/// Saves a `<artifact>.cargo-sbom.json` file for the given [`Unit`].
+///
+pub fn output_sbom(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> CargoResult<()> {
+    let bcx = build_runner.bcx;
+    let rustc: SbomRustc = bcx.rustc().into();
+
+    let packages = fetch_packages(build_runner, unit);
+
+    for sbom_output_file in build_runner.sbom_output_files(unit)? {
+        let sbom = Sbom::new(unit, packages.clone(), rustc.clone());
+
+        let mut outfile = BufWriter::new(paths::create(sbom_output_file)?);
+        let output = serde_json::to_string_pretty(&sbom)?;
```

Comment:
Perhaps pretty not important at this moment. Just wonder how big the size of the output sbom would be, and at some point we might want to use [`serde_json::to_writer_pretty`](https://docs.rs/serde_json/latest/serde_json/fn.to_writer_pretty.html) instead.

Reply:
As a rule of thumb, about the same as `cargo metadata`.

Since this is meant to be consumed by machines (and humans can easily pipe it to `jq` or open the file in a browser to inspect it), I think we should stick to minified output because this could be emitted and parsed on every release build, including incremental ones.

Reply:
> Since this is meant to be consumed by machines

Agree in that regard